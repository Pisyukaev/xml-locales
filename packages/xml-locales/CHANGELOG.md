# xml-locales

## 0.0.3

### Patch Changes

- refactor modifiers functions, utils and tests

## 0.0.2

### Patch Changes

- include dist folder to npm publish

## 0.0.1

### Patch Changes

- added repository urls in package.json
