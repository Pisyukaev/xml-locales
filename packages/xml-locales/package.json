{
	"name": "xml-locales",
	"description": "Tool for locales in xml files",
	"version": "0.0.3",
	"type": "module",
	"keywords": [
		"xml-locales",
		"xml",
		"locales"
	],
	"repository": {
		"type": "git",
		"url": "https://github.com/Pisyukaev/xml-locales.git",
		"directory": "packages/xml-locales"
	},
	"files": [
		"dist"
	],
	"types": "./dist/index.d.ts",
	"typesVersions": {
		"*": {
			"utils": [
				"./dist/utils/index.d.ts"
			]
		}
	},
	"exports": {
		".": "./dist/index.js",
		"./utils": "./dist/utils/index.js"
	},
	"scripts": {
		"dev": "tsup --watch",
		"build": "tsup",
		"test": "vitest run --coverage",
		"test:watch": "vitest --ui --watch --coverage"
	}
}
