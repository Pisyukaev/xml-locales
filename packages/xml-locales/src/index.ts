export * from './xml-locales.js';
export { XmlJsonData } from './utils/xml.js';
export type * from './utils/types.js';
