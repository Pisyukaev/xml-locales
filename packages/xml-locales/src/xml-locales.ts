import { addXmlNode } from './modifiers/add.js';
import {
	deleteByKeyXmlNode,
	deleteByValueXmlNode
} from './modifiers/delete.js';
import { sortXmlNodes } from './modifiers/sort.js';
import { updateXmlNode } from './modifiers/update.js';
import { XmlJsonData, XmlParser } from './utils/xml.js';
import type {
	AddOptions,
	SortDirection,
	UpdateOptions,
	XmlDataTypes
} from './utils/types.js';
import type { XmlConstructor } from './utils/xml.js';

export class XmlLocales {
	/** @internal */
	private readonly xmlParser: XmlParser;
	/** @internal */
	private xmlOptions?: XmlConstructor;
	/** @internal */
	private xmlData: XmlJsonData;

	constructor(
		xmlData?: string | Buffer | XmlJsonData,
		xmlOptions?: XmlConstructor
	) {
		this.xmlParser = new XmlParser(xmlOptions);
		this.xmlOptions = xmlOptions;
		this.parseXml(xmlData);
	}

	/** @internal */
	private parseXml(xmlData?: XmlDataTypes): void {
		if (xmlData instanceof XmlJsonData) {
			this.xmlData = xmlData;
			return;
		}

		if (xmlData) {
			const parsedXml = this.xmlParser.xmlToJson(xmlData);
			if (Array.isArray(parsedXml.resources.string)) {
				this.xmlData = parsedXml;
			} else {
				this.xmlData = new XmlJsonData({
					resources: { string: [parsedXml.resources.string] }
				});
			}
			return;
		}

		if (!this.xmlData) {
			throw new Error('XML data is not defined');
		}
	}

	/** @internal */
	private newInstance(): XmlLocales {
		return new XmlLocales(this.xmlData, this.xmlOptions);
	}

	add(options: AddOptions): XmlLocales {
		this.xmlData = addXmlNode(this.xmlData, options.key, options.value);
		return this.newInstance();
	}

	update(options: UpdateOptions): XmlLocales {
		this.xmlData = updateXmlNode(
			this.xmlData,
			options.oldValue,
			options.newValue
		);
		return this.newInstance();
	}

	deleteByKey(key: string): XmlLocales {
		this.xmlData = deleteByKeyXmlNode(this.xmlData, key);
		return this.newInstance();
	}

	deleteByValue(value: string): XmlLocales {
		this.xmlData = deleteByValueXmlNode(this.xmlData, value);
		return this.newInstance();
	}

	sort(sortDirection: SortDirection): XmlLocales {
		this.xmlData = sortXmlNodes(this.xmlData, sortDirection);
		return this.newInstance();
	}

	toJSON(): XmlJsonData {
		return this.xmlData;
	}

	toXML(formate = true): string {
		const xmlString = this.xmlParser.jsonToXml(this.xmlData);
		if (formate) return this.xmlParser.formate(xmlString);
		return xmlString;
	}
}
