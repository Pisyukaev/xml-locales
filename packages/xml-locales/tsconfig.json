{
	"extends": "@pisyukaev/tsconfig",
	"compilerOptions": {
		"outDir": "dist",
		"noUnusedLocals": false,
		"noUnusedParameters": false
	},
	"include": ["src", "tests"]
}
