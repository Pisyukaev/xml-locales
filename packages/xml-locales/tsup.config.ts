import { defineConfig } from 'tsup';

export default defineConfig({
	entry: ['src/index.ts', 'src/utils/index.ts'],
	target: 'node20',
	format: ['esm'],
	banner: {
		js: `
// BANNER START
const require = (await import("node:module")).createRequire(import.meta.url);
const __filename = (await import("node:url")).fileURLToPath(import.meta.url);
const __dirname = (await import("node:path")).dirname(__filename);
// BANNER END
`
	},
	minify: false,
	sourcemap: true,
	clean: true,
	dts: true
});
