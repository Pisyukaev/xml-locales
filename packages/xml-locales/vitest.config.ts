import { defineProject } from 'vitest/config';

export default defineProject({
	test: {
		name: 'xml-locales',
		environment: 'node'
	}
});
