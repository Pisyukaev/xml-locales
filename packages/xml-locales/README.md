# xml-locales

[![NPM](https://img.shields.io/npm/v/xml-locales.svg)](https://www.npmjs.com/package/xml-locales)

This is core package to work with localization files in xml.

## Installation

```sh
npm install xml-locales
```

```sh
yarn add xml-locales
```

```sh
pnpm add xml-locales
```

> [!IMPORTANT]
> This package, which works with XML files, has one root node named `resources`. This root node has child nodes named `string`. For example:
>
> ```xml
> <resources>
>   <string name="key1">value1</string>
>   <string name="key2">value2</string>
> </resources> 
>```

## Properties

| Args | Type | Required | Description |
|------|------|----------|-------------|
| `xmlData` | `string` \| `Buffer` \| `XmlJsonData` | `false` | Data of the xml document |
| `xmlOptions` | `object` | `false` | It represents the options for the XML parser.

### xmlOptions

The `xmlOptions` is used to customize the behavior of the XML parser and builder in the `XmlLocales` class.

| Prop | Type | Required | Description |
|------|------|----------|-------------|
| `parserOptions` | `X2jOptionsOptional` | `false` | Options to customize how the XML data is parsed. See below for details. For more information, see [here](https://github.com/NaturalIntelligence/fast-xml-parser/blob/master/docs/v4/2.XMLparseOptions.md) |
| `builderOptions`| `XmlBuilderOptionsOptional` | `false` | Options to customize how the XML data is built. See below for details. For more information, see [here](https://github.com/NaturalIntelligence/fast-xml-parser/blob/master/docs/v4/3.XMLBuilder.md) |
| `formateOptions`| `XMLFormatterOptions` | `false` | Options to customize the formatting of the XML data. See below for details. For more information, see [here](https://github.com/chrisbottin/xml-formatter#options) |

### Default Parser Options (X2jOptionsOptional)

| Option                | Value  |
|-----------------------|--------|
| `trimValues`          | `false`  |
| `ignoreDeclaration`   | `true`   |
| `attributeNamePrefix` | `'key_'` |
| `alwaysCreateTextNode`| `true`   |
| `ignoreAttributes`    | `false`  |

### Default Builder Options (XmlBuilderOptionsOptional)

| Option                | Value  |
|-----------------------|--------|
| `ignoreAttributes`    | `false` |
| `attributeNamePrefix` | `'key_'` |
| `processEntities`     | `false` |

### Default Formatter Options (XMLFormatterOptions)

| Option           | Value  |
|------------------|--------|
| `collapseContent`| `true` |
| `indentation`    | `'  '` |

## Usage

```js
import {XmlLocales} from 'xml-locales'

const xmlData = `
<resources>
  <string name="key1">value1</string>
  <string name="key2">value2</string>
</resources>
`

const xmlLocales = new XmlLocales(xmlData)
const jsonData = xmlLocales.add({key: 'newKey', value: 'newValue'}).toXML()

console.log(jsonData)
```

Output:

```xml
<resources>
   <string name="key1">value1</string>
   <string name="key2">value2</string>
   <string name="newKey">newValue</string>
</resources>
```

### Update key/value

```js
import {XmlLocales} from 'xml-locales'

const xmlData = `
<resources>
  <string name="key1">value1</string>
  <string name="key2">value2</string>
</resources>
`

const xmlLocales = new XmlLocales(xmlData)
const jsonData = xmlLocales.add({key: 'newKey', value: 'newValue'})
.update({oldValue: 'key1', newValue: 'firstKey'})
.update({oldValue: 'value2', newValue: 'secondValue'})
.toXML()

console.log(jsonData)
```

Output:

```xml
<resources>
   <string name="firstKey">value1</string>
   <string name="key2">secondValue</string>
</resources>
```

### Delete by key/value

```js
import {XmlLocales} from 'xml-locales'

const xmlData = `
<resources>
  <string name="key1">value1</string>
  <string name="key2">value2</string>
</resources>
`

const xmlLocales = new XmlLocales(xmlData)
const jsonData = xmlLocales.add({key: 'newKey', value: 'newValue'})
.deleteByKey('key1')
.deleteByValue('value2')
.toXML()

console.log(jsonData)
```

Output:

```xml
<resources>
</resources>
```

### XML and JSON

```js
import {XmlLocales} from 'xml-locales'

const xmlData = `
<resources>
  <string name="key1">value1</string>
  <string name="key2">value2</string>
</resources>
`

const xmlLocales = new XmlLocales(xmlData)
const xmlString = xmlLocales.toXML()
const jsonXml = xmlLocales.toJSON()

console.log(xmlString) // output XML
console.log(jsonXml) // output JSON
```

Output XML:

```xml
<resources>
  <string name="newKey2">newValue2</string>
  <string name="newKey">newValue</string>
  <string name="key2">value2</string>
  <string name="firstKey">value1</string>
</resources>
 ```

Output JSON:

```json
{
 resources: {
  string: [
   {
    key_name: 'key1',
    '#text': 'value1'
   },
   {
    key_name: 'key2',
    '#text': 'value2'
   }
  ]
 }
}
```

### Chaining

```js
import {XmlLocales} from 'xml-locales'

const xmlData = `
<resources>
  <string name="key1">value1</string>
  <string name="key2">value2</string>
</resources>
`

const xmlLocales = new XmlLocales(xmlData)

const jsonData = xmlLocales.add({key: 'newKey', value: 'newValue'})
.add({key: 'newKey2', value: 'newValue2'})
.update({oldValue: 'key1', newValue: 'firstKey'})
.sort('desc')
.toXML()

console.log(jsonData)
```

Output:

```xml
<resources>
  <string name="newKey2">newValue2</string>
  <string name="newKey">newValue</string>
  <string name="key2">value2</string>
  <string name="firstKey">value1</string>
</resources>
```

## Packages

| Package  | version   |
| ------- | -------- |
| ðŸ“¦[xml-locales](https://github.com/Pisyukaev/xml-locales/tree/master/packages/xml-locales) | [![NPM](https://img.shields.io/npm/v/xml-locales.svg)](https://www.npmjs.com/package/xml-locales)   |
| ðŸ’»[@xml-locales/cli](https://github.com/Pisyukaev/xml-locales/tree/master/packages/cli)   | [![NPM](https://img.shields.io/npm/v/@xml-locales/cli.svg)](https://www.npmjs.com/package/@xml-locales/cli)  |
