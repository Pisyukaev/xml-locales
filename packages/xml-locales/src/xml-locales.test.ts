import { describe, expect, test } from 'vitest';
import { c } from 'vitest/dist/reporters-5f784f42.js';

import { XmlLocales } from './xml-locales.js';

const xmlData = `<resources>
  <string name="key1">value1</string>
  <string name="key2">value2</string>
</resources>`;

const oneStringXmlData = `<resources>
  <string name="key1">value1</string>
</resources>`;

describe('XmlLocales', () => {
	describe('add', () => {
		test('should add new node', () => {
			const xmlLocales = new XmlLocales(xmlData);
			const jsonXml = xmlLocales.add({ key: 'key3', value: 'value3' }).toJSON();
			const newNode = jsonXml.resources.string[2];

			expect(newNode?.key_name).toBe('key3');
			expect(newNode?.['#text']).toBe('value3');
		});

		test('should replace the value of the key', () => {
			const xmlLocales = new XmlLocales(xmlData);
			const jsonXml = xmlLocales.add({ key: 'key1', value: 'value3' }).toJSON();
			const newNode = jsonXml.resources.string[0];

			expect(newNode?.key_name).toBe('key1');
			expect(newNode?.['#text']).toBe('value3');
		});
	});

	describe('remove', () => {
		test('should remove a node of the key', () => {
			const xmlLocales = new XmlLocales(xmlData);
			xmlLocales.deleteByKey('key1');
			const jsonXml = xmlLocales.toJSON();

			expect(jsonXml.resources.string[0]?.['#text']).toBe('value2');
		});

		test('should remove a node of the value', () => {
			const xmlLocales = new XmlLocales(xmlData);
			xmlLocales.deleteByValue('value1');
			const jsonXml = xmlLocales.toJSON();

			expect(jsonXml.resources.string[0]?.['#text']).toBe('value2');
		});

		test('not found key or value', () => {
			const xmlLocales = new XmlLocales(xmlData);
			xmlLocales.deleteByKey('key3');
			const jsonXml = xmlLocales.toJSON();

			expect(jsonXml.resources.string[0]?.['#text']).toBe('value1');
		});
	});

	describe('sort', () => {
		test('should sorted by "asc"', () => {
			const xmlLocales = new XmlLocales(xmlData);
			xmlLocales.sort('asc');
			const jsonXml = xmlLocales.toJSON();

			expect(jsonXml.resources.string[0]?.key_name).toBe('key1');
		});

		test('should sorted by "desc"', () => {
			const xmlLocales = new XmlLocales(xmlData);
			xmlLocales.sort('desc');
			const jsonXml = xmlLocales.toJSON();

			expect(jsonXml.resources.string[1]?.key_name).toBe('key1');
		});
	});

	describe('update', () => {
		test('should update key', () => {
			const xmlLocales = new XmlLocales(xmlData);
			xmlLocales.update({ oldValue: 'key1', newValue: 'newKey1' });
			const jsonXml = xmlLocales.toJSON();

			expect(jsonXml.resources.string[0]?.key_name).toBe('newKey1');
		});

		test('should update value', () => {
			const xmlLocales = new XmlLocales(xmlData);
			xmlLocales.update({ oldValue: 'value1', newValue: 'newValue1' });
			const jsonXml = xmlLocales.toJSON();

			expect(jsonXml.resources.string[0]?.['#text']).toBe('newValue1');
		});
	});

	describe('XmlLocales constructor', () => {
		test('should throw error if xml data is not defined', () => {
			expect(() => new XmlLocales()).toThrow('XML data is not defined');
		});

		test('constructor should parse xml data', () => {
			const xmlLocales = new XmlLocales(oneStringXmlData);
			const jsonXml = xmlLocales.toJSON();

			expect(jsonXml).toMatchSnapshot();
		});

		test('should return json object', () => {
			const xmlLocales = new XmlLocales(xmlData);
			const jsonXml = xmlLocales.toJSON();

			expect(jsonXml).toMatchSnapshot();
		});
	});

	describe('to XML', () => {
		test('should return xml string with formatting', () => {
			const xmlLocales = new XmlLocales(xmlData);
			const xmlString = xmlLocales.toXML();

			expect(xmlString).toMatchSnapshot();
		});

		test('should return xml string without formatting', () => {
			const xmlLocales = new XmlLocales(xmlData);
			const xmlString = xmlLocales.toXML(false);

			expect(xmlString).toMatchSnapshot();
		});
	});
});
