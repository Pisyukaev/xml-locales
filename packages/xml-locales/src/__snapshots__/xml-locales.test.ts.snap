// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`XmlLocales > XmlLocales constructor > constructor should parse xml data 1`] = `
XmlJsonData {
  "resources": {
    "string": [
      {
        "#text": "value1",
        "key_name": "key1",
      },
    ],
  },
}
`;

exports[`XmlLocales > XmlLocales constructor > should return json object 1`] = `
XmlJsonData {
  "resources": {
    "#text": "
  
  
",
    "string": [
      {
        "#text": "value1",
        "key_name": "key1",
      },
      {
        "#text": "value2",
        "key_name": "key2",
      },
    ],
  },
}
`;

exports[`XmlLocales > should return json object 1`] = `
XmlJsonData {
  "resources": {
    "#text": "
  
  
",
    "string": [
      {
        "#text": "value1",
        "key_name": "key1",
      },
      {
        "#text": "value2",
        "key_name": "key2",
      },
    ],
  },
}
`;

exports[`XmlLocales > should return xml string 1`] = `
"<resources>
  <string name=\\"key1\\">value1</string>
  <string name=\\"key2\\">value2</string>
</resources>"
`;

exports[`XmlLocales > to XML > should return xml string with formatting 1`] = `
"<resources>
  <string name=\\"key1\\">value1</string>
  <string name=\\"key2\\">value2</string>
</resources>"
`;

exports[`XmlLocales > to XML > should return xml string without formatting 1`] = `
"<resources><string name=\\"key1\\">value1</string><string name=\\"key2\\">value2</string>
  
  
</resources>"
`;
